README.md
# dcos-traefik

[Traefik](https://traefik.io) package for DC/OS.

`traefik.toml` configuration file is generated using `confd` template. Currently configuration can be loaded from ENV variables, evetually support for consistent key-value storages like Etcd, ZooKeeper, Consul can be added.

## Quickstart

In order to expose a service from Marathon add labels to your app:

```
labels: {
  "traefik.enable": "true",
  "traefik.frontend.rule": "Host:site.example.com",
  "traefik.protocol": "http",
  "traefik.backend.healthcheck.path": "/health",
  "traefik.frontend.entryPoints":"http,https"
}
```
Then you should be able to reach you site via Traefik's HTTP port (`TRAEFIK_HTTP_PORT`):
```
curl -H 'Host:site.example.com' localhost:80/health
```

See [documetation](https://docs.traefik.io/user-guide/marathon/) for more details.

## Global configuration

 * `TRAEFIK_DEBUG` Default `false`
 * `TRAEFIK_INSECURE_SKIP` Default `false`
 * `TRAEFIK_LOG_LEVEL` Default `INFO`
 * `TRAEFIK_DEFAULT_ENTRYPOINTS` Entrypoints to be used by frontends that do not specify any entrypoint. Each frontend can specify its own entrypoints.
 * `TRAEFIK_GRACE_TIMEOUT` Duration to give active requests a chance to finish before Traefik stops (`10s`)
 * `TRAEFIK_HEALTHCHECK_INTERVAL` Set the default health check interval (`30s`).
 * `TRAEFIK_READ_TIMEOUT` The maximum duration for reading the entire request, including the body.
 * `TRAEFIK_WRITE_TIMEOUT` The maximum duration before timing out writes of the response.
 * `TRAEFIK_IDLE_TIMEOUT` The maximum duration an idle (keep-alive) connection will remain idle before closing itself.
 * `TRAEFIK_RETRY` Number of attempts to reach backend service.

## Logs

### Access logs

By default the Traefik log is written to stdout in text format.

* `TRAEFIK_ACCESS_LOG` Enable logging. Default `false`
* `TRAEFIK_ACCESS_FORMAT` Logs format, e.g. `json`
* `TRAEFIK_ACCESS_PATH` Path to log files.

## Entrypoints

### http

* `TRAEFIK_HTTP_COMPRESSION` Default `true`
* `TRAEFIK_HTTP_ADDRESS`
* `TRAEFIK_HTTP_PORT` Default `80`

### https

 * `TRAEFIK_HTTPS_COMPRESSION` Default `true`
 * `TRAEFIK_HTTPS_ENABLE` Default `false`
 * `TRAEFIK_HTTPS_ADDRESS`
 * `TRAEFIK_HTTPS_PORT` Default `443`
 * `TRAEFIK_HTTPS_CERT` Certificate file, e.g. `examples/traefik.crt`
 * `TRAEFIK_HTTPS_KEY` Certificate key, e.g. `examples/traefik.key`

### ping

Heath check endpoint, responds without authentication to `/ping`.

 * `TRAEFIK_PING_ENABLE` Default `true`
 * `TRAEFIK_PING_PORT` If variable not defined, `$PORT0` will be used.

### API

Former `web.admin` interface.

 * `TRAEFIK_API_ENABLE` Enable API entrypoint
 * `TRAEFIK_API_PORT` If variable not defined, `$PORT1` will be used.
 * `TRAEFIK_API_DASHBOARD` Default `true`
 * `TRAEFIK_API_DEBUG` This will install HTTP handlers to expose Go expvars under /debug/vars and pprof profiling. Default `false`.

## Custom configuration

Appends custom configuration to generated `traefik.toml` config.

 * `TRAEFIK_FILE_NAME` Default `rules.toml`
 * `TRAEFIK_FILE_WATCH` Default `true`

## Providers

### Marathon

 * `TRAEFIK_MARATHON_ENABLE` Default `false`
 * `TRAEFIK_MARATHON_ENDPOINT` `http://marathon.mesos:8080`
 * `TRAEFIK_MARATHON_WATCH` Default `true`
 * `TRAEFIK_MARATHON_LB_COMPATIBILITY` Default `false`
 * `TRAEFIK_MARATHON_DOMAIN` Default `marathon.localhost`
 * `TRAEFIK_MARATHON_EXPOSE` Expose Marathon apps by default in Traefik. Default `true`
 * `TRAEFIK_MARATHON_GROUPS_AS_SUBDOMAINS` Convert Marathon groups to subdomains. Default `false`
 * `TRAEFIK_MARATHON_DIALER_TIMEOUT` Default `60s`
 * `TRAEFIK_MARATHON_KEEP_ALIVE` Default `10s`
 * `TRAEFIK_MARATHON_FORCE_TASK_HOSTNAME` Default `false`
 * `TRAEFIK_MARATHON_RESPECT_READINESS_CHECKS` Default `false`

### Mesos

 * `TRAEFIK_MESOS_ENABLE` Default `false`
 * `TRAEFIK_MESOS_ENDPOINT` `http://leader.mesos:5050`
 * `TRAEFIK_MESOS_WATCH` Default `true`
 * `TRAEFIK_MESOS_EXPOSE` Expose Mesos apps by default in Traefik. Default `false`
 * `TRAEFIK_MESOS_FILENAME` Override default configuration template.
 * `TRAEFIK_MESOS_ZK_TIMEOUT` ZooKeeper timeout (in seconds).
 * `TRAEFIK_MESOS_REFRESH` Polling interval (in seconds).
 * `TRAEFIK_MESOS_IP_SOURCES` IP sources (e.g. host, docker, mesos, netinfo).
 * `TRAEFIK_MESOS_TIMEOUT` HTTP Timeout (in seconds).
 * `TRAEFIK_MESOS_GROUPS_AS_SUBDOMAINS` Convert Mesos groups to subdomains. Default `false`

### Kubernetess

 * `TRAEFIK_K8S_ENABLE` Default `false`
 * `TRAEFIK_K8S_ENDPOINT` Kubernetes server endpoint.
 * `TRAEFIK_K8S_TOKEN` Bearer token used for the Kubernetes client configuration.
 * `TRAEFIK_K8S_CA` Path to the certificate authority file.
 * `TRAEFIK_K8S_NAMESPACES` Comma-separated namespaces to watch. e.g. `"\"default\",\"production\""`
 * `TRAEFIK_K8S_LABELSELECTOR` Ingress label selector to filter Ingress objects that should be processed. e.g. `"A and not B"`
 * `TRAEFIK_K8S_FILENAME` Override default configuration template.
 * `TRAEFIK_K8S_INGRESS_CLASS` Value of `kubernetes.io/ingress.class` annotation that identifies Ingress objects to be processed.
 * `TRAEFIK_K8S_DISABLE_PASS_HOST_HEADERS` Disable PassHost Headers.
 * `TRAEFIK_K8S_ENABLE_PASS_TLS_CERT` Enable PassTLSCert Headers.

### Rancher

 * `TRAEFIK_RANCHER_ENABLE` Default `false`
 * `TRAEFIK_RANCHER_REFRESH` Default `15`
 * `TRAEFIK_RANCHER_MODE` Default `api`
 * `TRAEFIK_RANCHER_DOMAIN` Default `rancher.localhost`
 * `TRAEFIK_RANCHER_EXPOSE` Default `true`
 * `TRAEFIK_RANCHER_FILENAME` Configuration template.
 * `TRAEFIK_RANCHER_HEALTHCHECK` Filter services with unhealthy states and inactive states. Default `true`
 * `TRAEFIK_RANCHER_INTERVALPOLL` Default `false`
 * `TRAEFIK_RANCHER_PREFIX` Default `/latest`

## Let's Encrypt

 * `TRAEFIK_ACME_ENABLE` Certificates from Let's Encrypt. Default `false`
 * `TRAEFIK_ACME_EMAIL`
 * `TRAEFIK_ACME_ENTRYPOINT` Default `https`.
 * `TRAEFIK_ACME_STORAGE` File or key used for certificates storage.
 * `TRAEFIK_ACME_ONHOSTRULE` Default `false`
 * `TRAEFIK_ACME_CASERVER` Default `https://acme-v01.api.letsencrypt.org/directory`

## Metrics

* `TRAEFIK_STATISTICS_RECENT_ERRORS` Number of recent errors logged. Default `10`.

### DataDog

* `TRAEFIK_DATADOG_ADDRESS` Enables sending metrics to DataDog, e.g. `localhost:8125`
* `TRAEFIK_DATADOG_PUSHINTERVAL` Default `10s`

### InfluxDB

* `TRAEFIK_INFLUXDB_ADDRESS` Enables sending metrics to InfluxDB, e.g. `localhost:8089`
* `TRAEFIK_INFLUXDB_PUSHINTERVAL` Default `10s`

### Prometheus

Will be enabled when `TRAEFIK_PROMETHEUS_ENTRYPOINT` is set, e.g. to `api`.

 * `TRAEFIK_PROMETHEUS_ENTRYPOINT`
 * `TRAEFIK_PROMETHEUS_BUCKETS` Comma separated values. Default `0.1,0.3,1.2,5.0`

### StatsD

* `TRAEFIK_STATSD_ADDRESS` Enables sending metrics to StatsD, e.g. `localhost:8125`
* `TRAEFIK_STATSD_PUSHINTERVAL` Default `10s`


## Tracing

* `TRAEFIK_TRACING_BACKEND` Backend name used to send tracing data, either `jaeger` or `zipkin`
* `TRAEFIK_TRACING_SERVICE` Service name used in tracing backend, default: `traefik`
* `TRAEFIK_TRACING_SPANLIMIT` Span name limit allows for name truncation in case of very long Frontend/Backend names

### Jaeger

* `TRAEFIK_TRACING_JAEGER_SERVER` Sampling Server URL is the address of jaeger-agent's HTTP sampling server
* `TRAEFIK_TRACING_JAEGER_SAMPLINGTYPE` Sampling Type specifies the type of the sampler: `const`, `probabilistic`, `rateLimiting`
* `TRAEFIK_TRACING_JAEGER_SAMPLINGPARAM` Sampling Param is a value passed to the sampler
* `TRAEFIK_TRACING_JAEGER_AGENTADDRESS` Local Agent Host Port instructs reporter to send spans to jaeger-agent at this address

### Zipkin

* `TRAEFIK_TRACING_ZIPKING_ENDPOINT` Zipking HTTP endpoint used to send data
* `TRAEFIK_TRACING_ZIPKING_DEBUG` Enable Zipkin debug, default: `false`
* `TRAEFIK_TRACING_ZIPKING_SAMESPAN` Use ZipKin SameSpan RPC style traces
* `TRAEFIK_TRACING_ZIPKING_ID128` Use ZipKin 128 bit root span IDs

CHANGELOG.md
# dcos-traefik Changelog

## v1.2.0

  - Fixed https endpoint (added TLS)
  - Added tracing support

## v1.1.0
  - Traefik configuration generated from `confd` template
  - Targets Traefik 1.6.x release
  - Rename `TRAEFIK_ADMIN_*` variables to `TRAEFIK_API_*`, `TRAEFIK_PING_*` see [documentation](https://docs.traefik.io/configuration/backends/web/)
  - Renamed `TRAEFIK_MARATHONLB_COMPATIBILITY` -> `TRAEFIK_MARATHON_LB_COMPATIBILITY`
  - Renamed `TRAEFIK_ADMIN_ENABLE` -> `TRAEFIK_API_ENABLE`
  - Renamed `TRAEFIK_ADMIN_PORT` -> `TRAEFIK_API_PORT`
  - Renamed `TRAEFIK_RANCHER_EXPOSED` -> `TRAEFIK_RANCHER_EXPOSE`
  - Renamed `TRAEFIK_ADMIN_STATISTICS` -> `TRAEFIK_STATISTICS_RECENT_ERRORS`
  - Removed `TRAEFIK_PROMETHEUS_ENABLE`, setting value to `TRAEFIK_PROMETHEUS_ENTRYPOINT` is enough.
  - Removed deprecated directives:
    * `TRAEFIK_ACME_ONDEMAND`
    * `TRAEFIK_ADMIN_READ_ONLY`
    * `TRAEFIK_ADMIN_AUTH_METHOD`
    * `TRAEFIK_ADMIN_AUTH_USERS`
  - Added Mesos provider
  - Improved Kubernetess support
  - Added Metrics support (exporting to DataDog, InfluxDB, StatsD)

## v1.0.0
 - Initial release
 - Targets Traefik 1.5.x release


bootstrap.sh
#!/bin/bash
export LC_ALL=C
#
function -h {
  cat <<USAGE
   USAGE: Generates Traefik config
   -b / --backend confd backend (configuration source)
   -d / --debug   debugging output
USAGE
}; function --help { -h ;}

function msg { out "$*" >&1 ;}
function out { printf '%s\n' "$*" ;}
function err { local x=$? ; msg "$*" ; return $(( $x == 0 ? 1 : $x )) ;}

function main {
  local debug=false
  # confd backend (default: ENV variables)
  local backend="env"
  while [[ $# -gt 0 ]]
  do
    case "$1" in                                      # Munging globals, beware
      -b|--backend)       backend=true; shift 1 ;;
      -d|--debug)         debug=true; shift 1 ;;
      *)                    err 'Argument error. Please see help: -h' ;;
    esac
  done
  if [[ $debug == true ]]; then
    set -ex
  fi
  if [ -f traefik_linux-amd64 ]; then
    chmod +x traefik_linux-amd64
  fi
  confd=$(ls -1 confd-*)
  if [ -f "${confd}" ]; then
    chmod +x ${confd}
  fi
  mkdir -p $(pwd)/{conf.d,templates}
  msg "Generating Traefik configuration"
  ./${confd} -onetime -backend ${backend} --confdir $(pwd)
}

if [[ ${1:-} ]] && declare -F | cut -d' ' -f3 | fgrep -qx -- "${1:-}"
then
  case "$1" in
    -h|--help) : ;;
    *) ;;
  esac
  "$@"
else
  main "$@"
fi


boostrap.sh
#!/bin/bash
export LC_ALL=C
#
function -h {
  cat <<USAGE
   USAGE: Generates Traefik config
   -v / --verbose  debugging output
USAGE
}; function --help { -h ;}

function msg { out "$*" >&1 ;}
function out { printf '%s\n' "$*" ;}
function err { local x=$? ; msg "$*" ; return $(( $x == 0 ? 1 : $x )) ;}

function main {
  local verbose=false
  while [[ $# -gt 0 ]]
  do
    case "$1" in                                      # Munging globals, beware
      -v|--verbose)         verbose=true; shift 1 ;;
      *)                    err 'Argument error. Please see help: -h' ;;
    esac
  done
  if [[ $verbose == true ]]; then
    set -ex
  fi
  generate_config
}

function generate_config {
  TRAEFIK_HTTP_COMPRESSION=${TRAEFIK_HTTP_COMPRESSION:-"true"}
  TRAEFIK_HTTPS_COMPRESSION=${TRAEFIK_HTTPS_COMPRESSION:-"true"}
  TRAEFIK_HTTP_ADDRESS=${TRAEFIK_HTTP_ADDRESS:-""}
  TRAEFIK_HTTP_PORT=${TRAEFIK_HTTP_PORT:-"8080"}
  TRAEFIK_HTTPS_ENABLE=${TRAEFIK_HTTPS_ENABLE:-"false"}
  TRAEFIK_HTTPS_ADDRESS=${TRAEFIK_HTTP_ADDRESS:-""}
  TRAEFIK_HTTPS_PORT=${TRAEFIK_HTTPS_PORT:-"8443"}
  TRAEFIK_ADMIN_ENABLE=${TRAEFIK_ADMIN_ENABLE:-"false"}
  TRAEFIK_ADMIN_PORT=${TRAEFIK_ADMIN_PORT:-"8000"}
  TRAEFIK_DEBUG=${TRAEFIK_DEBUG:="false"}
  TRAEFIK_INSECURE_SKIP=${TRAEFIK_INSECURE_SKIP:="false"}
  TRAEFIK_LOG_LEVEL=${TRAEFIK_LOG_LEVEL:-"INFO"}
  TRAEFIK_ADMIN_READ_ONLY=${TRAEFIK_ADMIN_READ_ONLY:="false"}
  TRAEFIK_ADMIN_STATISTICS=${TRAEFIK_ADMIN_STATISTICS:-10}
  TRAEFIK_ADMIN_AUTH_METHOD=${TRAEFIK_ADMIN_AUTH_METHOD:-"basic"}
  TRAEFIK_ADMIN_AUTH_USERS=${TRAEFIK_ADMIN_AUTH_USERS:-""}
  TRAEFIK_SSL_PATH=${TRAEFIK_SSL_PATH:-"$(pwd)/certs"}
  TRAEFIK_ACME_ENABLE=${TRAEFIK_ACME_ENABLE:-"false"}
  TRAEFIK_ACME_EMAIL=${TRAEFIK_ACME_EMAIL:-"test@traefik.io"}
  TRAEFIK_ACME_ONDEMAND=${TRAEFIK_ACME_ONDEMAND:-"true"}
  TRAEFIK_ACME_ONHOSTRULE=${TRAEFIK_ACME_ONHOSTRULE:-"true"}
  TRAEFIK_ACME_CASERVER=${TRAEFIK_ACME_CASERVER:-"https://acme-v01.api.letsencrypt.org/directory"}
  TRAEFIK_K8S_ENABLE=${TRAEFIK_K8S_ENABLE:-"false"}
  TRAEFIK_K8S_OPTS=${TRAEFIK_K8S_OPTS:-""}
  TRAEFIK_PROMETHEUS_ENABLE=${TRAEFIK_PROMETHEUS_ENABLE:-"false"}
  TRAEFIK_PROMETHEUS_OPTS=${TRAEFIK_PROMETHEUS_OPTS:-""}
  TRAEFIK_PROMETHEUS_BUCKETS=${TRAEFIK_PROMETHEUS_BUCKETS:-"[0.1,0.3,1.2,5.0]"}
  TRAEFIK_RANCHER_ENABLE=${TRAEFIK_RANCHER_ENABLE:-"false"}
  TRAEFIK_RANCHER_REFRESH=${TRAEFIK_RANCHER_REFRESH:-15}
  TRAEFIK_RANCHER_MODE=${TRAEFIK_RANCHER_MODE:-"api"}
  TRAEFIK_RANCHER_DOMAIN=${TRAEFIK_RANCHER_DOMAIN:-"rancher.internal"}
  TRAEFIK_RANCHER_EXPOSED=${TRAEFIK_RANCHER_EXPOSED:-"false"}
  TRAEFIK_RANCHER_HEALTHCHECK=${TRAEFIK_RANCHER_HEALTHCHECK:-"true"}
  TRAEFIK_RANCHER_INTERVALPOLL=${TRAEFIK_RANCHER_INTERVALPOLL:-"false"}
  TRAEFIK_RANCHER_OPTS=${TRAEFIK_RANCHER_OPTS:-""}
  TRAEFIK_RANCHER_PREFIX=${TRAEFIK_RANCHER_PREFIX:-"/2017-11-11"}
  TRAEFIK_FILE_NAME=${TRAEFIK_FILE_NAME:-"rules.toml"}
  TRAEFIK_FILE_ENABLE=${TRAEFIK_FILE_ENABLE:="false"}
  TRAEFIK_FILE_OPTS=${TRAEFIK_FILE_OPTS:-""}
  TRAEFIK_WEB=${TRAEFIK_WEB:-""}
  CATTLE_URL=${CATTLE_URL:-""}
  CATTLE_ACCESS_KEY=${CATTLE_ACCESS_KEY:-""}
  CATTLE_SECRET_KEY=${CATTLE_SECRET_KEY:-""}
  TRAEFIK_MARATHON_ENABLE=${TRAEFIK_MARATHON_ENABLE:-"true"}
  TRAEFIK_MARATHON_ENDPOINT=${TRAEFIK_MARATHON_ENDPOINT:-"http://marathon.mesos:8080"}
  TRAEFIK_MARATHON_WATCH=${TRAEFIK_MARATHON_WATCH:-"true"}
  TRAEFIK_MARATHON_DOMAIN=${TRAEFIK_MARATHON_DOMAIN:-"marathon.localhost"}

TRAEFIK_ENTRYPOINTS_HTTP="\
  [entryPoints.http]
  address = \"${TRAEFIK_HTTP_ADDRESS}:${TRAEFIK_HTTP_PORT}\"
  compress = ${TRAEFIK_HTTP_COMPRESSION}
"

if [ -d "${TRAEFIK_SSL_PATH}" ]; then
  filelist=`ls -1 ${TRAEFIK_SSL_PATH}/*.key | rev | cut -d"." -f2- | rev`
  RC=`echo $?`

  if [ $RC -eq 0 ]; then
    TRAEFIK_ENTRYPOINTS_HTTPS="\
  [entryPoints.https]
  address = \"${TRAEFIK_HTTPS_ADDRESS}:${TRAEFIK_HTTPS_PORT}\"
  compress = ${TRAEFIK_HTTPS_COMPRESSION}
    [entryPoints.https.tls]"
    for i in $filelist; do
        if [ -f "$i.crt" ]; then
            TRAEFIK_ENTRYPOINTS_HTTPS=$TRAEFIK_ENTRYPOINTS_HTTPS"
      [[entryPoints.https.tls.certificates]]
      certFile = \"${i}.crt\"
      keyFile = \"${i}.key\"
"
        fi
      done
  fi
fi

if [ "X${TRAEFIK_HTTPS_ENABLE}" == "Xtrue" ]; then
    TRAEFIK_ENTRYPOINTS_OPTS=${TRAEFIK_ENTRYPOINTS_HTTP}${TRAEFIK_ENTRYPOINTS_HTTPS}
    TRAEFIK_ENTRYPOINTS='"http", "https"'
elif [ "X${TRAEFIK_HTTPS_ENABLE}" == "Xonly" ]; then
    TRAEFIK_ENTRYPOINTS_HTTP=$TRAEFIK_ENTRYPOINTS_HTTP"\
    [entryPoints.http.redirect]
       entryPoint = \"https\"
"
    TRAEFIK_ENTRYPOINTS_OPTS=${TRAEFIK_ENTRYPOINTS_HTTP}${TRAEFIK_ENTRYPOINTS_HTTPS}
    TRAEFIK_ENTRYPOINTS='"http", "https"'
else
    TRAEFIK_ENTRYPOINTS_OPTS=${TRAEFIK_ENTRYPOINTS_HTTP}
    TRAEFIK_ENTRYPOINTS='"http"'
fi

if [ "X${TRAEFIK_K8S_ENABLE}" == "Xtrue" ]; then
    TRAEFIK_K8S_OPTS="[kubernetes]"
fi

TRAEFIK_ACME_CFG=""
if [ "X${TRAEFIK_HTTPS_ENABLE}" == "Xtrue" ] || [ "X${TRAEFIK_HTTPS_ENABLE}" == "Xonly" ] && [ "X${TRAEFIK_ACME_ENABLE}" == "Xtrue" ]; then

    TRAEFIK_ACME_CFG="\
[acme]
email = \"${TRAEFIK_ACME_EMAIL}\"
storage = \"${SERVICE_HOME}/acme/acme.json\"
onDemand = ${TRAEFIK_ACME_ONDEMAND}
OnHostRule = ${TRAEFIK_ACME_ONHOSTRULE}
caServer = \"${TRAEFIK_ACME_CASERVER}\"
entryPoint = \"https\"
"

fi

if [ "X${TRAEFIK_RANCHER_ENABLE}" == "Xtrue" ]; then
    TRAEFIK_RANCHER_OPTS="\
[rancher]
domain = \"${TRAEFIK_RANCHER_DOMAIN}\"
Watch = true
RefreshSeconds = ${TRAEFIK_RANCHER_REFRESH}
ExposedByDefault = ${TRAEFIK_RANCHER_EXPOSED}
EnableServiceHealthFilter = ${TRAEFIK_RANCHER_HEALTHCHECK}
"
    if [ "${TRAEFIK_RANCHER_MODE}" == "api" ]; then
        TRAEFIK_RANCHER_OPTS=${TRAEFIK_RANCHER_OPTS}"
[rancher.api]
Endpoint = \"${CATTLE_URL}\"
AccessKey = \"${CATTLE_ACCESS_KEY}\"
SecretKey = \"${CATTLE_SECRET_KEY}\"
"
    elif [ "${TRAEFIK_RANCHER_MODE}" == "metadata" ]; then
        TRAEFIK_RANCHER_OPTS=${TRAEFIK_RANCHER_OPTS}"
[rancher.metadata]
IntervalPoll = ${TRAEFIK_RANCHER_INTERVALPOLL}
Prefix = \"${TRAEFIK_RANCHER_PREFIX}\"
"
    fi
fi

if [ "X${TRAEFIK_ADMIN_ENABLE}" == "Xtrue" ] || [ "X${TRAEFIK_PROMETHEUS_ENABLE}" == "Xtrue" ]; then
    TRAEFIK_WEB="\
[web]"
fi

if [ "X${TRAEFIK_ADMIN_ENABLE}" == "Xtrue" ]; then
    TRAEFIK_ADMIN_CFG="\
address = \":${TRAEFIK_ADMIN_PORT}\"
ReadOnly = ${TRAEFIK_ADMIN_READ_ONLY}
[web.statistics]
RecentErrors = ${TRAEFIK_ADMIN_STATISTICS}
"

    if [ "${TRAEFIK_ADMIN_AUTH_USERS}" != "" ]; then
        echo ${TRAEFIK_ADMIN_AUTH_USERS} > "${SERVICE_HOME}/.htpasswd"
        TRAEFIK_ADMIN_CFG=${TRAEFIK_ADMIN_CFG}"
[web.auth.${TRAEFIK_ADMIN_AUTH_METHOD}]
usersFile = \"${SERVICE_HOME}/.htpasswd\"
"
    fi
fi

if [ "X${TRAEFIK_PROMETHEUS_ENABLE}" == "Xtrue" ]; then
    TRAEFIK_PROMETHEUS_OPTS="\
[web.metrics.prometheus]
buckets=${TRAEFIK_PROMETHEUS_BUCKETS}
"
fi

if [ "X${TRAEFIK_FILE_ENABLE}" == "Xtrue" ]; then
    TRAEFIK_FILE_OPTS="\
[file]
filename = \"${TRAEFIK_FILE_NAME}\"
watch = true
"
fi

if [ "X${TRAEFIK_MARATHON_ENABLE}" == "Xtrue" ]; then
    TRAEFIK_MARATHON_OPTS="\
[marathon]
# Marathon server endpoint.
# You can also specify multiple endpoint for Marathon:
# endpoint = \"http://10.241.1.71:8080,10.241.1.72:8080,10.241.1.73:8080\"
#
endpoint = \"${TRAEFIK_MARATHON_ENDPOINT}\"
watch = ${TRAEFIK_MARATHON_WATCH}

# Default domain used.
# Can be overridden by setting the \"traefik.domain\" label on an application.
#
# Required
#
domain = \"${TRAEFIK_MARATHON_DOMAIN}\"

# Enable compatibility with marathon-lb labels.
#
# Optional
# Default: false
marathonLBCompatibility = ${TRAEFIK_MARATHONLB_COMPATIBILITY}

# Expose Marathon apps by default in Traefik.
#
# Optional
# Default: true
#
exposedByDefault = ${TRAEFIK_MARATHON_EXPOSE}
"
fi



cat << EOF > ./traefik.toml
# traefik.toml
debug = ${TRAEFIK_DEBUG}
logLevel = "${TRAEFIK_LOG_LEVEL}"
InsecureSkipVerify = ${TRAEFIK_INSECURE_SKIP}
defaultEntryPoints = [${TRAEFIK_ENTRYPOINTS}]
[entryPoints]
${TRAEFIK_ENTRYPOINTS_OPTS}
${TRAEFIK_WEB}
${TRAEFIK_ADMIN_CFG}
${TRAEFIK_PROMETHEUS_OPTS}
${TRAEFIK_RANCHER_OPTS}
${TRAEFIK_FILE_OPTS}
${TRAEFIK_ACME_CFG}
${TRAEFIK_K8S_OPTS}
EOF
}


if [[ ${1:-} ]] && declare -F | cut -d' ' -f3 | fgrep -qx -- "${1:-}"
then
  case "$1" in
    -h|--help) : ;;
    *) ;;
  esac
  "$@"
else
  main "$@"
fi
